File: robotics-codes-from-scratch-master/python/iLQR_bimanual.py
'''
iLQR applied to a planar bimanual robot for a tracking problem involving
the center of mass (CoM) and the end-effector (batch formulation)

Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
Written by Sylvain Calinon <https://calinon.ch>

This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
License: MIT
'''

import numpy as np
#import numpy.matlib
import matplotlib.pyplot as plt
import matplotlib.patches as patches

# Helper functions
# ===============================

# Forward kinematics for end-effector (in robot coordinate system)
def fkin(x, param):
    L = np.tril(np.ones(3))
    f = np.vstack([
        param.l[0:3].T @ np.cos(L @ x[0:3]),
        param.l[0:3].T @ np.sin(L @ x[0:3]),
        param.l[[0,3,4]].T @ np.cos(L @ x[[0,3,4]]),
        param.l[[0,3,4]].T @ np.sin(L @ x[[0,3,4]])
    ])  # f1,f2,f3,f4
    return f

# Forward kinematics for end-effector (in robot coordinate system)
def fkin0(x, param): 
    L = np.tril(np.ones(3))
    fl = np.vstack([
        L @ np.diag(param.l[0:3]) @ np.cos(L @ x[0:3]),
        L @ np.diag(param.l[0:3]) @ np.sin(L @ x[0:3])
    ])
    fr = np.vstack([
        L @ np.diag(param.l[[0,3,4]]) @ np.cos(L @ x[[0,3,4]]),
        L @ np.diag(param.l[[0,3,4]]) @ np.sin(L @ x[[0,3,4]])
    ])
    f = np.hstack([fl[:,::-1], np.zeros([2,1]), fr])
    return f

# Jacobian of the end-effector with analytical computation (for single time step)
def Jkin(x, param):
    L = np.tril(np.ones(3))
    J = np.zeros((param.nbVarF, param.nbVarX))
    Jl = np.vstack([-np.sin(L @ x[:3]).T @ np.diag(param.l[:3]) @ L,
                    np.cos(L @ x[:3]).T @ np.diag(param.l[:3]) @ L
                    ])
    Jr = np.vstack([-np.sin(L @ x[[0,3,4]]).T @ np.diag(np.array(param.l)[[0,3,4]]) @ L,
                    np.cos(L @ x[[0,3,4]]).T @ np.diag(np.array(param.l)[[0,3,4]]) @ L
                    ])
    J[:Jl.shape[0], :Jl.shape[1]] = Jl
    J[2:, [0,3,4]] = Jr
    return J

# Residual and Jacobian of end-effector for a viapoints reaching task (in robot coordinate system)
def f_reach(x, param):
    f = fkin(x, param) - param.Mu
    J = np.zeros((param.nbVarF*param.nbPoints, param.nbVarX*param.nbPoints))
    for t in range(param.nbPoints):
        Jtmp = Jkin(x[:,t], param)
        J[t * param.nbVarF:(t + 1) * param.nbVarF, t * param.nbVarX:(t + 1) * param.nbVarX] = Jtmp
    return f, J
        
# Forward kinematics of the center of mass
def fkin_CoM(x, param):
    L = np.tril(np.ones(3))
    f = np.vstack([param.l[:3] @ L @ np.cos(L @ x[:3,:]) +
                   np.array(param.l)[[0,3,4]] @ L @ np.cos(L @ x[[0,3,4],:]),
                   param.l[:3] @ L @ np.sin(L @ x[:3,:]) +
                   np.array(param.l)[[0,3,4]] @ L @ np.sin(L @ x[[0,3,4],:])
                   ]) / 6
    return f

# Jacobian  of the center of mass with analytical computation (for single time step)
def Jkin_CoM(x, param):
    L = np.tril(np.ones(3))
    Jl = np.vstack([-np.sin(L @ x[:3]).T @ L @ np.diag(param.l[:3].T @ L),
                    np.cos(L @ x[:3]).T @ L @ np.diag(param.l[:3].T @ L)
                    ]) / 6
    Jr = np.vstack([-np.sin(L @ x[[0,3,4]]).T @ L @ np.diag(np.array(param.l)[[0,3,4]].T @ L),
                    np.cos(L @ x[[0,3,4]]).T @ L @ np.diag(np.array(param.l)[[0,3,4]].T @ L)
                    ]) / 6
    #J = np.hstack(((Jl[:,0] + Jr[:,0]).reshape(-1,1), Jl[:,1:], Jr[:,1:]))
    J = np.hstack([(Jl[:,0] + Jr[:,0])[:,np.newaxis], Jl[:,1:], Jr[:,1:]])
    return J

# Residual and Jacobian of Center of Mass for a viapoints reaching task (in robot coordinate system)
def f_reach_CoM(x, param):
    #f = fkin_CoM(x, param) -np.matlib.repmat(param.MuCoM, 1, param.nbData)
    f = fkin_CoM(x, param) - param.MuCoM
    J = np.zeros([2*param.nbData, param.nbVarX*param.nbData])
    for t in range(param.nbData):
        Jtmp = Jkin_CoM(x[:,t], param)
        J[t*2:(t+1)*2, t*param.nbVarX:(t+1)*param.nbVarX] = Jtmp
    return f, J


param = lambda: None # Lazy way to define an empty class in python
param.dt = 1e0 # Time step length
param.nbData = 30 # Number of datapoints
param.nbIter = 100 # Maximum number of iterations for iLQR
param.nbPoints = 1 # Number of viapoints
param.nbVarX = 5 # State space dimension ([q1,q2,q3] for left arm, [q1,q4,q5] for right arm)
param.nbVarU = param.nbVarX # Control space dimension (dq1,dq2,dq3,dq4,dq5)
param.nbVarF = 4 # Task space dimension ([x1,x2] for left end-effector, [x3,x4] for right end-effector)
param.l = np.ones(param.nbVarX) * 2 # Robot links lengths
    
param.r = 1e-5 # Control weight term
param.Mu = np.array([[-1, -1.5, 4, 2]]).T # Target point for end-effectors
param.MuCoM = np.array([[0, 1.4]]).T # Target point for center of mass

# Main program
# ===============================

# Control weight matrix (at trajectory level)
R = np.eye(param.nbVarU * (param.nbData-1)) * param.r
# Precision matrix for end-effectors tracking
Q = np.kron(np.eye(param.nbPoints), np.diag([1, 1, 0, 0]))
# Precision matrix for continuous CoM tracking
Qc = np.kron(np.eye(param.nbData), np.diag([1, 1]))

# Time occurence of viapoints
tl = np.linspace(0, param.nbData-1, param.nbPoints+1 )
tl = np.fromiter(map(lambda x: int(x + 0.5), tl[1:]), dtype=np.int32)
idx = (tl)[:,np.newaxis] * param.nbVarX + np.arange(param.nbVarU)

# initial setup
u = np.zeros(param.nbVarU * (param.nbData-1)) # Initial control command
x0 = np.array([np.pi/2, np.pi/2, np.pi/3, -np.pi/2, -np.pi/3])#Initial pose

# Solving Iterative LQR (iLQR)
# ===============================

# Transfer matrices (for linear system as single integrator)
Su0 = np.vstack([np.zeros((param.nbVarX, param.nbVarX*(param.nbData-1))),
      np.tril(np.kron(np.ones((param.nbData-1, param.nbData-1)), np.eye(param.nbVarX)*param.dt))])
Sx0 = np.kron( np.ones(param.nbData) , np.identity(param.nbVarX) ).T
Su = Su0[idx.flatten()] # We remove the lines that are out of interest

for i in range(param.nbIter):
    x = Su0 @ u + Sx0 @ x0  # System evolution
    x = x.reshape([param.nbVarX, param.nbData], order='F')
    f, J = f_reach(x[:,tl], param)  # Forward kinematics and Jacobian for end-effectors
    fc, Jc = f_reach_CoM(x, param)  # Forward kinematics and Jacobian for center of mass
#    print(Qc.shape)
#    print(fc.flatten('F'))

    du = np.linalg.inv(Su.T @ J.T @ Q @ J @ Su + Su0.T @ Jc.T @ Qc @ Jc @ Su0 + R) @ \
         (-Su.T @ J.T @ Q @ f.flatten('F') - \
         Su0.T @ Jc.T @ Qc @ fc.flatten('F') - \
         u * param.r)

    # Estimate step size with line search method
    alpha = 1
    cost0 = f.flatten('F').T @ Q @ f.flatten('F') + np.linalg.norm(u)**2 * param.r  # Cost
    while True:
        utmp = u + du * alpha
        xtmp = Su0 @ utmp + Sx0 @ x0  # System evolution
        xtmp = xtmp.reshape([param.nbVarX, param.nbData], order='F')
        ftmp, _ = f_reach(xtmp[:,tl], param)
        fctmp, _ = f_reach_CoM(xtmp, param)

        # for end-effectors and CoM
        cost = ftmp.flatten('F').T @ Q @ ftmp.flatten('F') + np.linalg.norm(utmp)**2 * param.r  # Cost
        if cost < cost0 or alpha < 1e-3:
            print("Iteration {}, cost: {}, alpha: {}".format(i, cost, alpha))
            break # Stop iLQR when solution is reached

        alpha *= .5

    u = u + du * alpha
    if np.linalg.norm(du * alpha) < 1e-2:
        break

#  Plot state space
tl = np.array([0, tl.item()])

plt.figure(figsize=(15, 9))
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

# Plot bimanual robot
ftmp = fkin0(x[:,0], param)
plt.plot(ftmp[0], ftmp[1], c='black', linewidth=4, alpha=.2)

ftmp = fkin0(x[:,-1], param)
plt.plot(ftmp[0], ftmp[1], c='black', linewidth=4, alpha=.6)

# Plot CoM
fc = fkin_CoM(x, param)  # Forward kinematics for center of mass
plt.plot(fc[0,0], fc[1,0], c='black', marker='o', linewidth=0,
         markersize=np.sqrt(90), markerfacecolor='none', markeredgewidth=3, alpha=.4)  # Plot CoM
plt.plot(fc[0,-1], fc[1,-1], c='black', marker='o', linewidth=0,
         markersize=np.sqrt(90), markerfacecolor='none', markeredgewidth=3, alpha=.6)  # Plot CoM

# Plot end-effectors targets
for t in range(param.nbPoints):
    plt.plot(param.Mu[0,t], param.Mu[1,t], marker='o', c='red', markersize=14)

# Plot CoM target
plt.plot(param.MuCoM[0], param.MuCoM[1], c='red', marker='o', linewidth=0,
         markersize=np.sqrt(90), markerfacecolor='none', markeredgewidth=2, alpha=.8)

# Plot end-effectors paths
ftmp = fkin(x, param)
plt.plot(ftmp[0,:], ftmp[1,:], c="black", marker="o", markevery=[0] + tl.tolist())
plt.plot(ftmp[2,:], ftmp[3,:], c="black", marker="o", markevery=[0] + tl.tolist())
plt.show()

File: robotics-codes-from-scratch-master/matlab/iLQR_bimanual.m
%% iLQR applied to a planar bimanual robot for a tracking problem involving 
%% the center of mass (CoM) and the end-effector (batch formulation)
%% 
%% Copyright (c) 2023 Idiap Research Institute <https://www.idiap.ch/>
%% Written by Sylvain Calinon <https://calinon.ch>
%% 
%% This file is part of RCFS <https://robotics-codes-from-scratch.github.io/>
%% License: MIT

function iLQR_bimanual

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.dt = 1E0; %Time step size
param.nbIter = 100; %Maximum number of iterations for iLQR
param.nbPoints = 1; %Number of viapoints
param.nbData = 30; %Number of datapoints
param.nbVarX = 5; %State space dimension ([q1,q2,q3] for left arm, [q1,q4,q5] for right arm)
param.nbVarU = param.nbVarX; %Control space dimension (dq1,dq2,dq3,dq4,dq5)
param.nbVarF = 4; %Task space dimension ([x1,x2] for left end-effector, [x3,x4] for right end-effector)
param.l = ones(param.nbVarX,1) * 2; %Robot links lengths
param.r = 1E-5; %Control weighting term
param.Mu = [-1; -1.5; 4; 2]; %Target point for end-effectors
%param.Mu = [-3; 1.5; 4.5; 3]; %Target point for end-effectors
param.MuCoM = [0; 1.4]; %Target point for center of mass

R = speye(param.nbVarU * (param.nbData-1)) * param.r; %Control weight matrix (at trajectory level)
Q = kron(speye(param.nbPoints), diag([1, 1, 0, 0])); %Precision matrix for end-effectors tracking 
Qc = kron(speye(param.nbData), diag([1, 1])); %Precision matrix for continuous CoM tracking 

%Time occurrence of viapoints
tl = linspace(1, param.nbData, param.nbPoints+1);
tl = round(tl(2:end));
idx = (tl - 1) * param.nbVarX + [1:param.nbVarX]';


%% Iterative LQR (iLQR)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
u = zeros(param.nbVarU*(param.nbData-1), 1);
x0 = [pi/2; pi/2; pi/3; -pi/2; -pi/3]; %Initial pose

%Transfer matrices (for linear system as single integrator)
Su0 = [zeros(param.nbVarX, param.nbVarX*(param.nbData-1)); tril(kron(ones(param.nbData-1), eye(param.nbVarX)*param.dt))];
Sx0 = kron(ones(param.nbData,1), eye(param.nbVarX));
Su = Su0(idx,:);

for n=1:param.nbIter	
	x = reshape(Su0 * u + Sx0 * x0, param.nbVarX, param.nbData); %System evolution
	[f, J] = f_reach(x(:,tl), param); %Forward kinematics and Jacobian for end-effectors
	[fc, Jc] = f_reach_CoM(x, param); %Forward kinematics and Jacobian for center of mass	
	du = (Su' * J' * Q * J * Su + Su0' * Jc' * Qc * Jc * Su0 + R) \ (-Su' * J' * Q * f - Su0' * Jc' * Qc * fc - u * param.r); %Gradient
	
	%Estimate step size with line search method
	alpha = 1;
	cost0 = f' * Q * f + fc' * Qc * fc + norm(u)^2 * param.r; %for end-effectors and CoM
	while 1
		utmp = u + du * alpha;
		xtmp = reshape(Su0 * utmp + Sx0 * x0, param.nbVarX, param.nbData);
		ftmp = f_reach(xtmp(:,tl), param);
		fctmp = f_reach_CoM(xtmp, param);
		cost = ftmp' * Q * ftmp + fctmp' * Qc * fctmp + norm(utmp)^2 * param.r; %for end-effectors and CoM
		if cost < cost0 || alpha < 1E-3
			break;
		end
		alpha = alpha * 0.5;
	end
	u = u + du * alpha; 
	
	if norm(du * alpha) < 1E-2
		break; %Stop iLQR when solution is reached
	end
end
disp(['iLQR converged in ' num2str(n) ' iterations.']);


%% Plot state space
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
tl = [1, tl];
colMat = lines(param.nbPoints*2);

h = figure('position',[10,10,1200,800],'color',[1,1,1]); hold on; axis off;
%Plot bimanual robot
ftmp = fkin0(x(:,1), param);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.8 .8 .8]);
ftmp = fkin0(x(:,tl(end)), param);
plot(ftmp(1,:), ftmp(2,:), '-','linewidth',4,'color',[.4 .4 .4]);
%Plot CoM
fc = fkin_CoM(x, param); %Forward kinematics for center of mass 
plot(fc(1,1), fc(2,1), 'o','linewidth',4,'markersize',12,'color',[.5 .5 .5]); %Plot CoM
plot(fc(1,tl(end)), fc(2,tl(end)), 'o','linewidth',4,'markersize',12,'color',[.2 .2 .2]); %Plot CoM
%Plot end-effectors targets
for t=1:param.nbPoints
	plot(param.Mu(1,t), param.Mu(2,t), '.','markersize',50,'color',colMat((t-1)*2+2,:));
	%plot(param.Mu(3,t), param.Mu(4,t), '.','markersize',50,'color',colMat((t-1)*2+1,:));
end
%Plot CoM target
plot(param.MuCoM(1,:), param.MuCoM(2,:), 'ro','linewidth',4,'markersize',12); %Plot end-effector target
%Plot end-effectors paths
ftmp = fkin(x, param);
plot(ftmp(1,:), ftmp(2,:), 'k-','linewidth',1);
plot(ftmp(3,:), ftmp(4,:), 'k-','linewidth',1);
plot(ftmp(1,tl), ftmp(2,tl), 'k.','markersize',20);
plot(ftmp(3,tl), ftmp(4,tl), 'k.','markersize',20);
axis equal; 

waitfor(h);
end 


%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for all articulations of a bimanual robot (in robot coordinate system)
function f = fkin0(x, param)
	L = tril(ones(3));
	fl = [L * diag(param.l(1:3)) * cos(L * x(1:3)), ...
	      L * diag(param.l(1:3)) * sin(L * x(1:3))]'; 
	fr = [L * diag(param.l([1,4:5])) * cos(L * x([1,4:5])), ...
	      L * diag(param.l([1,4:5])) * sin(L * x([1,4:5]))]';
	f = [fliplr(fl), zeros(2,1), fr];
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for bimanual robot (in robot coordinate system)
function f = fkin(x, param)
	L = tril(ones(3));
	f = [param.l(1:3)' * cos(L * x(1:3,:)); ...
	     param.l(1:3)' * sin(L * x(1:3,:)); ...
	     param.l([1,4:5])' * cos(L * x([1,4:5],:)); ...
	     param.l([1,4:5])' * sin(L * x([1,4:5],:))];
end

%%%%%%%%%%%%%%%%%%%%%%
%Forward kinematics for center of mass of a bimanual robot (in robot coordinate system)
function f = fkin_CoM(x, param)
	L = tril(ones(3));
	f = [param.l(1:3)' * L * cos(L * x(1:3,:)) + param.l([1,4:5])' * L * cos(L * x([1,4:5],:)); ...
	     param.l(1:3)' * L * sin(L * x(1:3,:)) + param.l([1,4:5])' * L * sin(L * x([1,4:5],:))] / 6;
end

%%%%%%%%%%%%%%%%%%%%%%
%Jacobian for bimanual robot (in robot coordinate system)
function J = Jkin(x, param)
	L = tril(ones(3));
	J = [-sin(L * x([1,2,3]))' * diag(param.l([1,2,3])) * L; ...
	      cos(L * x([1,2,3]))' * diag(param.l([1,2,3])) * L];
	J(3:4,[1,4:5]) = [-sin(L * x([1,4,5]))' * diag(param.l([1,4,5])) * L; ...
	                   cos(L * x([1,4,5]))' * diag(param.l([1,4,5])) * L];
end

%%%%%%%%%%%%%%%%%%%%%%
%Jacobian for center of mass of a bimanual robot (in robot coordinate system)
function J = Jkin_CoM(x, param)
	L = tril(ones(3));
	Jl = [-sin(L * x(1:3))' * L * diag(param.l(1:3)' * L); ...
	       cos(L * x(1:3))' * L * diag(param.l(1:3)' * L)] / 6;
	Jr = [-sin(L * x([1,4:5]))' * L * diag(param.l([1,4:5])' * L); ...
	       cos(L * x([1,4:5]))' * L * diag(param.l([1,4:5])' * L)] / 6;
	J = [(Jl(:,1) + Jr(:,1)), Jl(:,2:end), Jr(:,2:end)]; %Jacobian for center of mass
end

%%%%%%%%%%%%%%%%%%%%%%
%Cost and gradient for end-effectors
function [f, J] = f_reach(x, param)
	f = fkin(x, param) - param.Mu;
	f = f(:);
	
	J = []; 
	for t=1:size(x,2)
		J = blkdiag(J, Jkin(x(:,t), param));
	end
end

%%%%%%%%%%%%%%%%%%%%%%
%Cost and gradient for center of mass of bimanual robot
function [f, J] = f_reach_CoM(x, param)
	f = fkin_CoM(x, param) - param.MuCoM; 
	f = f(:);

	J = []; 
	for t=1:size(x,2)
		J = blkdiag(J, Jkin_CoM(x(:,t), param));
	end
end
